Fix undefined behaviour for operators -= and |= on guards

fbrausse opened the pull request:
The `guardt` class defines a conjunction of formulas that are to be used as the antecedent of implications. I documented a few of its properties and came across the operator-=() on guards. It is implemented using [std::set_difference()](https://en.cppreference.com/w/cpp/algorithm/set_difference), which assumes that the two ranges given to it are sorted. This operator is used in phi_function().

As far as I can see in the implementation on master the vectors of expressions in a guard are not sorted. So this (draft) PR checks what happens when they are.

fbrausse committed:
[util] guard: documentation + potential bug?

fbrausse committed:
[util] guard: sort the ranges before applying set_difference()

fbrausse committed:
[util] guard: remove declarations of default constructors implicitly there

fbrausse committed:
[util] guard: use a std::set instead of always sorting a vector

mikhailramalho commented:
The SMT backend relies on the order of operands to cache the converted expressions, you'll see huge slowdowns if the order of the guards is not stable (we should document it somewhere...).

Also, we should probably make guardt an expr2tc too and see if we can save some memory in the process

fbrausse commented:
Thanks for the background info, @mikhailramalho. That's certainly useful. I'm not proposing to do this. But I've been seeing crashes locally when testing this. As I understand it, there should just be performance changes, no semantic differences.

Crashes are occurring here (as far as I can see): https://github.com/esbmc/esbmc/blob/441d3dcee5b58434dfc7a1e20914483c670a1004/src/goto-symex/execution_state.cpp#L550

mikhailramalho commented:
Em ter., 22 de ago. de 2023 às 19:41, Franz Brauße ***@***.***>
escreveu:

> Thanks for the background info, @mikhailramalho
> <https://github.com/mikhailramalho>. That's certainly useful. I'm not
> proposing to do this.
>
Sure, it was just a heads-up in case you get down the rabbit hole and start
seeing some regressions.


> But I've been seeing crashes results locally when testing this. As I
> understand it, there should just be performance changes, no semantic
> differences.
>
> Crashes are occurring here (as far as I can see):
> https://github.com/esbmc/esbmc/blob/441d3dcee5b58434dfc7a1e20914483c670a1004/src/goto-symex/execution_state.cpp#L550
>

Yeah, another piece of code that relies on guard ordering that we are
unaware of.


> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1297#issuecomment-1689018940>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH3R32UVTK5UXNPHNM3XWUYRNANCNFSM6AAAAAA32SIXEE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

fbrausse commented:
> Yeah, another piece of code that relies on guard ordering that we are unaware of.

Maybe, yes. But actually my intention was testing the operator-=() on guards. According to the library docs, std::set_difference() is undefined when the ranges are not sorted by operator<() on expressions. On master they are not.

fbrausse committed:
[regression] fixed known bug

fbrausse committed:
[util] guard: replace undefined std::set_difference() with manual loop

fbrausse committed:
[util] guard: replace manual loop over bin-search by hash-lookup

fbrausse committed:
Revert "[util] guard: use a std::set instead of always sorting a vector"

This reverts commit 78a3a3742d5da9f4f3c06f7b5a2edf87bfb5cb56.

fbrausse committed:
[util, smt] fix pointer arith on --32 for 64-bit constants

Fixes regression/esbmc/github_243

fbrausse committed:
[regression] mark known bug: regression/mathsat/github_1070

MathSAT v5.5.4 fails on this example now when building the error trace
with

  ERROR: Error creating SMT
  ERROR: Error text: "Invalid model"

in mathsat_convt::get_bool() for this smt_astt:

  (set-info :source |printed by MathSAT|)
  (assert false)

Noteworthy, msat_get_model_value() already spews out a huge amount of
internal mathsat representation.

fbrausse committed:
[mathsat] dump() to log_status()

fbrausse commented:
Switching back to the unsorted vector, but fixing the documented undefined behaviour of std::set_difference() in operator-=() has semantic changes, even though it shouldn't - if I caught the indended meaning of operator-=() correctly. It shows a few issues that have been addressed in this PR, but which exist independent of it.

fbrausse committed:
[util] guard: mention order is important

fbrausse commented:
The usage of operator-=() in phi_function() is unchanged from CBMC. However, there they implement the operator not as set-difference, but as prefix removal.

fbrausse committed:
[util] guard: document & restore semantics of operator-=()

fbrausse commented:
> Also, we should probably make guardt an expr2tc too and see if we can save some memory in the process

@mikhailramalho I remember you mentioned this before; could you explain how that should work? Would it be an expr2tc of boolean type? To me it sounds a bit strange, because conceptually for me it is this lambda-expression: given an expression 'e', produce 'guard implies e'. Therefore it should be of function type.

mikhailramalho commented:
yes, it would have type bool and would hold a vector<expr2tc>, much like
the structs/unions in esbmc.

AFAIK, whenever we want the expr2tc from the guard, we call as_expr()
anyway. We could get rid of this and handle expert::guardt down the line
(probably only at smt level?).

Em qua., 23 de ago. de 2023 às 09:32, Franz Brauße ***@***.***>
escreveu:

> Also, we should probably make guardt an expr2tc too and see if we can save
> some memory in the process
>
> @mikhailramalho <https://github.com/mikhailramalho> I remember you
> mentioned this before; could you explain how that should work? Would it be
> an expr2tc of boolean type? To me it sounds a bit strange, because
> conceptually for me it is this lambda-expression: given an expression 'e',
> produce 'guard implies e'. Therefore it should have type 'expr2tc ->
> expr2tc'.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1297#issuecomment-1689879292>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZ43NHDKPHBYZ2HWITXWXZ4TANCNFSM6AAAAAA32SIXEE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

fbrausse committed:
[cmake] allow mathsat version number components to be multiple digits

fbrausse committed:
[regression] known bug: esbmc-unix2/14_popl20-figure1

When moving the abort() call out of the atomic block, as is done in
esbmc-unix2/14_popl20-figure1-no-atomic-abort, the test is successful.

fbrausse commented:
@mikhailramalho Thanks, I'll think about it. Wouldn't be much of a difference to an arbitrary-size conjunction... I won't try to do it in this PR.

This PR is about documenting the known behaviour of guards as we have them right now. Current status:
- operator-=() on guards uses undefined functionality from the C++ std lib: std::set_difference assumes ranges are sorted. This is fixed in this PR by restoring the semantics CBMC has for this operator: removal of the common prefix.
- operator|=() on guards still uses C++ std lib's set operations, which assume sorted ranges. I will address this next.

Minor other fixes:
- fix an issue where the smt_memspace wrongly assumed pointer arithmetic would only be performed with operands of <= bits of that of the pointer
- allow MathSAT version components to have multiple digits, such as the current 5.6.10

fbrausse commented:
About the one new KNOWNBUG in this PR: I found the statement
> Notice that it is not possible to simply wrap the assertion in an atomic block because if the assertion is violated, the abort would redirect the control-flow outside of the atomic block, violating the svcomp rules.

in https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1396 which to me indicates that there are some rules in SV-COMP (where the KNOWNBUG regression test case is from) that disallow calling abort() inside an atomic block. So it may actually not be a bug, depending on what semantics ESBMC assigns to its atomic blocks. @lucasccordeiro @mikhailramalho Do you know anything about abort() inside atomic blocks?

fbrausse committed:
[util] guard: refactor operator|= to not use library set operations with undefined behaviour

fbrausse committed:
Merge remote-tracking branch 'upstream/master' into fb/guard-set

fbrausse committed:
[regression] repaired mathsat/github_1070

fbrausse committed:
[util] guard: doxy-fy comments

fbrausse committed:
[util] guard: fix class invariant for build_guard_expr()

The invariant is that g_expr is always equal to the conjunction of all
expressions in guard_list as soon as it's not empty.
guardt::build_guard_expr() violated this condition in case it was called
when the list only had a single entry.

fbrausse committed:
[util] guard: document class invariant

fbrausse committed:
[util] guard: make operators -= and |= operate on sets

fbrausse committed:
[symex] do not std::move from guards

The function merge_state_guards() was taken from CBMC in
ff829c0fd324e0f4322505450a56d68444ef203b. Contrary to them we
still use the guards after the call to merge_state_guards() so
move-constructing the return value from the either state will
clear that guard (i.e., make it true) and its subsequent usage
will be wrong.

This affects, for instance, the SV-COMP'23 unreach-call benchmark
- c/aws-c-common/aws_array_list_pop_back_harness.i
under #744.

fbrausse committed:
[util] guard: restore invariant also for make_true()

fbrausse committed:
[util] guard: rm unused common_prefix_size()

fbrausse committed:
[symex] merge state guards as a last step

fbrausse committed:
[regression] revert KNOWNBUG changes to esbmc-unix2/14_popl20-figure1 and esbmc/11_cars
